### ISSUE TITLE
[Bug] Amiga Core - Black screen after loading WHLoader ROM

### ISSUE THREAD

**Issue by shaylevary** [emulatorjs/emulatorjs#692]
Hi,

First of all thank you very much for all your effort and work, I really appreciate it.

I was looking forward for the Amiga Core to be added to your script, since I have been using the emulatorjs.com version to run Amiga games, but I always prefer to use your code which is more flexible and run the games much faster and compatible.

Unfortunately there is probably a problem with the new Amiga core code, after loading the files, the core loads and the emulator starts, but the game does not load and instead of it - a black screen appears (I've tested  several different games that work properly in the other script).

Attached is a log for your review.
[amiga_log.txt](https://github.com/EmulatorJS/EmulatorJS/files/12789788/amiga_log.txt)

**Comment by michael-j-green** [comment]
Silly question... have you tried different Kickstart ROM's? I notice you loaded the one for the A1200, maybe try a 1.3 or 2.0 for an A500?

**Comment by shaylevary** [comment]
I'm using Amiga kickstart bios package which worked for me ok with emulatorjs.com script, it always use to select the compatible kickstart for the specific game automatically.

I will try to search if i can find another version of kickstart for the A1200 model instead of this file kickstart (kick40068.A1200).


emulator.js:1047 [libretro INFO] Model: 'A1200'
emulator.js:1047 [libretro INFO] Kickstart: 'kick40068.A1200'

**Comment by ethanaobrien** [comment]
As a note, your firmware file name must be the exact same as mentioned here

https://docs.libretro.com/library/puae/

Please try collecting them all and supplying a zip if possible

**Comment by michael-j-green** [comment]
This is what I did... loaded all the supported kickstarts into a zip and supplied it to EJS.

Pang! loaded first go.

**Comment by shaylevary** [comment]
Can you please help me with bios package?

**Comment by michael-j-green** [comment]
Probably shouldn't ask for ROM's on the GitHub page ;)

**Comment by michael-j-green** [comment]
Come join the Discord and we can discuss the issue you're having.

**Comment by shaylevary** [comment]
Hi,

I would like to know if it would be possible to try to solve the problem of loading Amiga game files in WHDL format.

These are files adapted from diskette format to a hard disk format, which makes loading time much faster, compared to the ADF diskette version.

When I try to run a game file in WHDL format with the alternative encrypted version of the script, the game loads properly and quickly.
When I try to run the same file in this open source script version, the game loads, but a few seconds of loading I get a black screen.
On the console log  the following error message appears which may be related to the reason why files in WHDL format are not loaded:
[libretro INFO] WHDLoad.prefs '//WHDLoad.prefs' not found, attempting to create.

[Amiga WHDL console log.txt](https://github.com/EmulatorJS/EmulatorJS/files/14298388/Amiga.WHDL.console.log.txt)

**Comment by michael-j-green** [comment]
Hey, I'll try and take a look at that this weekend. Hoping it's just as simple as the WHDLoad.prefs file not existing.

**Comment by shaylevary** [comment]
Thank you for trying fixing this issue, I really appreciate it, it will change the loading time significantly.
I'm available anytime if you need any help with testing the WHDL game files.

**Comment by shaylevary** [comment]
> Hey, I'll try and take a look at that this weekend. Hoping it's just as simple as the WHDLoad.prefs file not existing.

Hi,

Did you have some time to check this issue with the amiga's whdl support ? I prefer using this open source script which works much faster, then the other encoded script which currently support whdl.

**Comment by michael-j-green** [comment]
Apologies; this is on me. Work and sick kid. Starting to check it out now.

**Comment by michael-j-green** [comment]
@shaylevary; I've taken a look at the WHDLoad issue. I found that the demo page strips the file of the extension. Unfortunately, PUAE uses the extension to determine if the file is a WHDLoad LHA or not. Sites that embed EJS don't seem to have this issue.

A work around is to zip the LHA file first so that when loaded, EJS will extract it with the correct filename.

As with many Amiga games though, you may need to mess with the model, and amount of RAM to get it to start. Go to https://docs.libretro.com/library/puae/#model-overriding and pick the model string you'd like to use, and add it to the end of the file name, but before the .lha. For example: ``ThisIsMyGame (A600), ECS.lha``

Let us know how you go.

**Comment by michael-j-green** [comment]
Ok, looked at it some more. I think we have a path issue here.

When I load ``BumpysArcadeFantasy_v1-0_1604.lha`` on a locally installed copy of RetroArch with the PUAE core installed, I get these mount messages:
```
[libretro INFO] Mounting uaehf.device:0 0 (0):
[libretro INFO] FS: mounted virtual unit WHDLoad (/Users/michaelgreen/Documents/RetroArch/saves/WHDLoad/)
[libretro INFO] Mounting uaehf.device:0 1 (0):
[libretro INFO] FS: mounted virtual unit WHDSaves (/Users/michaelgreen/Documents/RetroArch/saves/WHDSaves/)
[libretro INFO] Mounting uaehf.device:0 2 (0):
[libretro INFO] FS: mounted virtual unit DH0 (/Users/michaelgreen/Documents/RetroArch/saves/TEMP/BumpysArcadeFantasy_v1-0_1604.lha)
[libretro INFO] FS: WHDLoad (flags=00000002,E=0,ED=1,EF=0,native='/Users/michaelgreen/Documents/RetroArch/saves/WHDLoad/') starting..
[libretro INFO] FS: WHDSaves (flags=00000002,E=0,ED=1,EF=0,native='/Users/michaelgreen/Documents/RetroArch/saves/WHDSaves/') starting..
[libretro INFO] FS: BumpysArcadeFantasy_v1-0_1604 (flags=00000004,E=0,ED=0,EF=1,native='/Users/michaelgreen/Documents/RetroArch/saves/TEMP/BumpysArcadeFantasy_v1-0_1604.lha') starting..
```

On the EJS demo page (supplying a zipped copy of the aforementioned lha file), I get this:
```
[Log] [libretro INFO] Mounting uaehf.device:0 0 (0): (emulator.js, line 1096)
[Log] [libretro INFO] FS: mounted virtual unit WHDLoad (/data/saves/PUAE/WHDLoad/) (emulator.js, line 1096)
[Log] [libretro INFO] Mounting uaehf.device:0 1 (0): (emulator.js, line 1096)
[Log] [libretro INFO] FS: mounted virtual unit WHDSaves (/data/saves/PUAE/WHDSaves/) (emulator.js, line 1096)
[Log] [libretro INFO] Mounting uaehf.device:0 2 (0): (emulator.js, line 1096)
[Log] [libretro INFO] FS: mounted virtual unit DH0 (/BumpysArcadeFantasy_v1-0_1604.lha) (emulator.js, line 1096)
[Log] [libretro INFO] FS: WHDLoad (flags=00000002,E=0,ED=1,EF=0,native='/data/saves/PUAE/WHDLoad/') starting.. (emulator.js, line 1096)
```

When I look in ``/data/saves/PUAE/WHDLoad/`` there are no contents (actually, nothing under ``/data/saves/``), so I believe it fails to start. It just sits there constantly checking the virtual HDD for the rest of the WHDLoad (press V, and select STBR to see HDD access in the bottom right corner).

**Comment by shaylevary** [comment]
> @shaylevary; I've taken a look at the WHDLoad issue. I found that the demo page strips the file of the extension. Unfortunately, PUAE uses the extension to determine if the file is a WHDLoad LHA or not. Sites that embed EJS don't seem to have this issue.
> 
> A work around is to zip the LHA file first so that when loaded, EJS will extract it with the correct filename.
> 
> As with many Amiga games though, you may need to mess with the model, and amount of RAM to get it to start. Go to https://docs.libretro.com/library/puae/#model-overriding and pick the model string you'd like to use, and add it to the end of the file name, but before the .lha. For example: `ThisIsMyGame (A600), ECS.lha`
> 
> Let us know how you go.

i've tried to rename the bumpy's lha file as you've suggested to this name:
BumpysArcadeFantasy (A600), ECS.lha

Then I've compressed it into zip file, and set ejs to load the zip file, but it still doesn't seems to work for me (i've attached my log).
[emulatorjs_amiga_log.txt](https://github.com/EmulatorJS/EmulatorJS/files/14395730/emulatorjs_amiga_log.txt)
Does this workaround work for you?

Also, regarding your conclusions: 
> When I look in /data/saves/PUAE/WHDLoad/ there are no contents (actually, nothing under /data/saves/), so I believe it fails to start. It just sits there constantly checking the virtual HDD for the rest of the WHDLoad (press V, and select STBR to see HDD access in the bottom right corner).

Is there any option to make EJS to support the WHDL on the same way it works on your local installed retroarch?

**Comment by michael-j-green** [comment]
I tried a number of different variations and couldn’t get it to work.

It looks like WHDLoad isn’t being properly setup at startup.

I’ve been experimenting with it this afternoon, but it’s almost bed time for my 2 month old, so it’ll have to wait until tomorrow before I can get back into it.

**Comment by shaylevary** [comment]
No problem, I appreciate your help.

**Comment by michael-j-green** [comment]
Stumbled upon something interesting... accidentally started EJS with the RetroArch.cfg disabled and PUAE extracted the WHDLoad files... no idea why it only works when no RetroArch.cfg is present, but there it is:
```
[Log] [libretro INFO] WHDLoad.prefs '//WHDLoad.prefs' not found, attempting to create.. (emulator.js, line 1096)
[Log] [libretro INFO] GUnzip: //WHDLoad.prefs (emulator.js, line 1096)
[Log] [libretro INFO] WHDLoad image directory '/home/web_user/retroarch/userdata/saves/PUAE/WHDLoad' not found, attempting to create.. (emulator.js, line 1096)
[Log] [libretro INFO] Mkdir: /home/web_user/retroarch/userdata/saves/PUAE/WHDLoad/C/ (emulator.js, line 1096)
[Log] [libretro INFO] Unzip: /home/web_user/retroarch/userdata/saves/PUAE/WHDLoad/C/Assign (emulator.js, line 1096)
[Log] [libretro INFO] Unzip: /home/web_user/retroarch/userdata/saves/PUAE/WHDLoad/C/ClickNot (emulator.js, line 1096)
[Log] [libretro INFO] Unzip: /home/web_user/retroarch/userdata/saves/PUAE/WHDLoad/C/Copy (emulator.js, line 1096)
[Log] [libretro INFO] Unzip: /home/web_user/retroarch/userdata/saves/PUAE/WHDLoad/C/Delete (emulator.js, line 1096)
```

**Comment by shaylevary** [comment]
First of all thank you for your help, I really appreciate.
Can you please tell me how do I set EJS to disable the retroarch.cfg file?

**Comment by shaylevary** [comment]
Hi,

Did you get any progress please in finding a solution for this issue, because loading Amiga games with standard disk settings takes ages as it emulate the old 3.5 diskette slow reading, while with WHLOADER HD it loads the game really fast.

Thank you,

**Comment by shaylevary** [comment]
Hi @michael-j-green ,

I would like to ask please if the new update "Disable CUE generation for the PUAE core" will also solve the problem with loading Amiga hd image file?

Thank you.

**Comment by ethanaobrien** [comment]
Fixed in 4.1.1
